### PR TITLE
Only default Job fields when feature gates are enabled

### LINKED ISSUE #100187: Job suspend and completionMode fields should be left empty when features are disabled
When feature gates are disabled, fields should not be rendered when storing or when returning objects from API.

For this reason, feature gate should control defaulting and the fields should be pointers.

This was missed during api-review for `.spec.suspend` and `.spec.completionMode`

/sig apps

### PR DESCRIPTION
**Pull request by alculquicondor**
#### What type of PR is this?

/kind bug
/kind api-change

#### What this PR does / why we need it:

Only default Job fields when feature gates are enabled and use pointer for completionMode enum.
This avoids rendering the field when the feature gates are disabled.
/sig apps

#### Which issue(s) this PR fixes:

Fixes #100187

kubernetes/enhancements#2214
kubernetes/enhancements#2232

#### Does this PR introduce a user-facing change?

Features not released yet

```release-note
NONE
```

### CODE REVIEW

**Comment by alculquicondor** [comment]
FYI @adtac

File: pkg/registry/batch/job/strategy_test.go [hunk 1]
```diff
@@ -153,21 +153,10 @@ func testJobStrategy(t *testing.T) {
 		t.Errorf("Expected a validation error")
 	}
 
-	// Existing gated fields should be preserved
```
Comment (adtac) [inline pkg/registry/batch/job/strategy_test.go:156]: why was this removed?

File: pkg/registry/batch/job/strategy.go [hunk 2]
```diff
@@ -115,8 +114,8 @@ func (jobStrategy) PrepareForUpdate(ctx context.Context, obj, old runtime.Object
 		newJob.Spec.TTLSecondsAfterFinished = nil
 	}
 
-	if !utilfeature.DefaultFeatureGate.Enabled(features.IndexedJob) && oldJob.Spec.CompletionMode == batch.NonIndexedCompletion {
-		newJob.Spec.CompletionMode = batch.NonIndexedCompletion
+	if !utilfeature.DefaultFeatureGate.Enabled(features.IndexedJob) && oldJob.Spec.CompletionMode == nil {
+		newJob.Spec.CompletionMode = nil
```
Comment (adtac) [inline pkg/registry/batch/job/strategy.go:118]: this would allow updating from NonIndexed->Indexed even when the feature gate is disabled, no?

File: pkg/registry/batch/job/strategy_test.go (hunk 1 above)
Comment (alculquicondor) [reply pkg/registry/batch/job/strategy_test.go:156]: To keep the test simple. Other fields are already tested above.
It would be ideal to have a test table so that we don't have to repeat the checks, but kind of late to do such refactoring.

File: pkg/registry/batch/job/strategy.go (hunk 2 above)
Comment (alculquicondor) [reply pkg/registry/batch/job/strategy.go:118]: It would... but field is immutable (checked in `ValidateJobSpecUpdate`)

Also, the feature gate is, in a way, protecting the field to not be set, without caring about the value.

**Comment by liggitt** [comment]
you need to update the API compatibility fixtures, basically undoing the changes made in https://github.com/kubernetes/kubernetes/pull/98441

**Comment by liggitt** [comment]
The API change itself lgtm

**Review by adtac — COMMENTED** [review]
controller and test changes lgtm

File: pkg/registry/batch/job/strategy_test.go (hunk 1 above)
Comment (adtac) [reply pkg/registry/batch/job/strategy_test.go:156]: I don't know if removing the TTL test is a good idea though

Comment (liggitt) [reply pkg/registry/batch/job/strategy_test.go:156]: > I don't know if removing the TTL test is a good idea though

/hold
just saw this... agree that this should not change unrelated fields

Comment (alculquicondor) [reply pkg/registry/batch/job/strategy_test.go:156]: changes in TTL is already tested above

Comment (liggitt) [reply pkg/registry/batch/job/strategy_test.go:156]: ack

/hold cancel

Comment (adtac) [reply pkg/registry/batch/job/strategy_test.go:156]: not for PrepareForUpdate though

Comment (alculquicondor) [reply pkg/registry/batch/job/strategy_test.go:156]: https://github.com/kubernetes/kubernetes/blob/b345913c5eb794ded1a577db66e9f566eb8e5957/pkg/registry/batch/job/strategy_test.go#L134
